File: types-old/out.ts
// To parse this data:
//
//   import { Convert, CfguType, Cfgu, ConfigSchemaType, ConfigSchema, Set, Store, StoreQueryElement, Config } from "./file";
//
//   const cfguType = Convert.toCfguType(json);
//   const cfgu = Convert.toCfgu(json);
//   const configSchemaType = Convert.toConfigSchemaType(json);
//   const configSchema = Convert.toConfigSchema(json);
//   const cfguContents = Convert.toCfguContents(json);
//   const set = Convert.toSet(json);
//   const store = Convert.toStore(json);
//   const storeQueryElement = Convert.toStoreQueryElement(json);
//   const storeQuery = Convert.toStoreQuery(json);
//   const config = Convert.toConfig(json);
//   const storeContents = Convert.toStoreContents(json);
//
// These functions will throw an error if the JSON doesn't
// match the expected interface, even if the JSON is valid.

/**
 * A generic representation of a <schema>.cfgu.[json|yaml] file that contains ConfigSchema
 * records ([<key>: string]: ConfigSchema)
 */
export interface Cfgu {
    contents: string;
    name:     string;
    path:     string;
    type:     CfguType;
}

export enum CfguType {
    JSON = "json",
    YAML = "yaml",
}

export interface CfguContents {
    default?:     string;
    depends?:     string[];
    description?: string;
    pattern?:     string;
    required?:    boolean;
    template?:    string;
    type:         ConfigSchemaType;
}

export enum ConfigSchemaType {
    Base64 = "Base64",
    Boolean = "Boolean",
    Color = "Color",
    ConnectionString = "ConnectionString",
    Country = "Country",
    Currency = "Currency",
    Domain = "Domain",
    Email = "Email",
    Hex = "Hex",
    IPv4 = "IPv4",
    IPv6 = "IPv6",
    LatLong = "LatLong",
    Locale = "Locale",
    Md5 = "MD5",
    MobilePhone = "MobilePhone",
    Number = "Number",
    RegEx = "RegEx",
    SHA = "SHA",
    SemVer = "SemVer",
    String = "String",
    URL = "URL",
    UUID = "UUID",
}

/**
 * A generic representation of a ConfigSet that contains a path hierarchy to differentiate
 * Config values and enable inheritance
 */
export interface Set {
    hierarchy: string[];
    path:      string;
}

/**
 * A generic representation of a ConfigStore that contains Config records (Config[])
 */
export interface Store {
    scheme: string;
    uid:    string;
}

export interface StoreQuery {
    key:    string;
    schema: string;
    set:    string;
}

/**
 * A generic representation of a software configuration, aka Config
 */
export interface StoreContents {
    key:    string;
    schema: string;
    set:    string;
    value:  string;
}

// Converts JSON strings to/from your types
// and asserts the results of JSON.parse at runtime
export class Convert {
    public static toCfguType(json: string): CfguType {
        return cast(JSON.parse(json), r("CfguType"));
    }

    public static cfguTypeToJson(value: CfguType): string {
        return JSON.stringify(uncast(value, r("CfguType")), null, 2);
    }

    public static toCfgu(json: string): Cfgu {
        return cast(JSON.parse(json), r("Cfgu"));
    }

    public static cfguToJson(value: Cfgu): string {
        return JSON.stringify(uncast(value, r("Cfgu")), null, 2);
    }

    public static toConfigSchemaType(json: string): ConfigSchemaType {
        return cast(JSON.parse(json), r("ConfigSchemaType"));
    }

    public static configSchemaTypeToJson(value: ConfigSchemaType): string {
        return JSON.stringify(uncast(value, r("ConfigSchemaType")), null, 2);
    }

    public static toConfigSchema(json: string): CfguContents {
        return cast(JSON.parse(json), r("CfguContents"));
    }

    public static configSchemaToJson(value: CfguContents): string {
        return JSON.stringify(uncast(value, r("CfguContents")), null, 2);
    }

    public static toCfguContents(json: string): { [key: string]: CfguContents } {
        return cast(JSON.parse(json), m(r("CfguContents")));
    }

    public static cfguContentsToJson(value: { [key: string]: CfguContents }): string {
        return JSON.stringify(uncast(value, m(r("CfguContents"))), null, 2);
    }

    public static toSet(json: string): Set {
        return cast(JSON.parse(json), r("Set"));
    }

    public static setToJson(value: Set): string {
        return JSON.stringify(uncast(value, r("Set")), null, 2);
    }

    public static toStore(json: string): Store {
        return cast(JSON.parse(json), r("Store"));
    }

    public static storeToJson(value: Store): string {
        return JSON.stringify(uncast(value, r("Store")), null, 2);
    }

    public static toStoreQueryElement(json: string): StoreQuery {
        return cast(JSON.parse(json), r("StoreQuery"));
    }

    public static storeQueryElementToJson(value: StoreQuery): string {
        return JSON.stringify(uncast(value, r("StoreQuery")), null, 2);
    }

    public static toStoreQuery(json: string): StoreQuery[] {
        return cast(JSON.parse(json), a(r("StoreQuery")));
    }

    public static storeQueryToJson(value: StoreQuery[]): string {
        return JSON.stringify(uncast(value, a(r("StoreQuery"))), null, 2);
    }

    public static toConfig(json: string): StoreContents {
        return cast(JSON.parse(json), r("StoreContents"));
    }

    public static configToJson(value: StoreContents): string {
        return JSON.stringify(uncast(value, r("StoreContents")), null, 2);
    }

    public static toStoreContents(json: string): StoreContents[] {
        return cast(JSON.parse(json), a(r("StoreContents")));
    }

    public static storeContentsToJson(value: StoreContents[]): string {
        return JSON.stringify(uncast(value, a(r("StoreContents"))), null, 2);
    }
}

function invalidValue(typ: any, val: any, key: any = ''): never {
    if (key) {
        throw Error(`Invalid value for key "${key}". Expected type ${JSON.stringify(typ)} but got ${JSON.stringify(val)}`);
    }
    throw Error(`Invalid value ${JSON.stringify(val)} for type ${JSON.stringify(typ)}`, );
}

function jsonToJSProps(typ: any): any {
    if (typ.jsonToJS === undefined) {
        const map: any = {};
        typ.props.forEach((p: any) => map[p.json] = { key: p.js, typ: p.typ });
        typ.jsonToJS = map;
    }
    return typ.jsonToJS;
}

function jsToJSONProps(typ: any): any {
    if (typ.jsToJSON === undefined) {
        const map: any = {};
        typ.props.forEach((p: any) => map[p.js] = { key: p.json, typ: p.typ });
        typ.jsToJSON = map;
    }
    return typ.jsToJSON;
}

function transform(val: any, typ: any, getProps: any, key: any = ''): any {
    function transformPrimitive(typ: string, val: any): any {
        if (typeof typ === typeof val) return val;
        return invalidValue(typ, val, key);
    }

    function transformUnion(typs: any[], val: any): any {
        // val must validate against one typ in typs
        const l = typs.length;
        for (let i = 0; i < l; i++) {
            const typ = typs[i];
            try {
                return transform(val, typ, getProps);
            } catch (_) {}
        }
        return invalidValue(typs, val);
    }

    function transformEnum(cases: string[], val: any): any {
        if (cases.indexOf(val) !== -1) return val;
        return invalidValue(cases, val);
    }

    function transformArray(typ: any, val: any): any {
        // val must be an array with no invalid elements
        if (!Array.isArray(val)) return invalidValue("array", val);
        return val.map(el => transform(el, typ, getProps));
    }

    function transformDate(val: any): any {
        if (val === null) {
            return null;
        }
        const d = new Date(val);
        if (isNaN(d.valueOf())) {
            return invalidValue("Date", val);
        }
        return d;
    }

    function transformObject(props: { [k: string]: any }, additional: any, val: any): any {
        if (val === null || typeof val !== "object" || Array.isArray(val)) {
            return invalidValue("object", val);
        }
        const result: any = {};
        Object.getOwnPropertyNames(props).forEach(key => {
            const prop = props[key];
            const v = Object.prototype.hasOwnProperty.call(val, key) ? val[key] : undefined;
            result[prop.key] = transform(v, prop.typ, getProps, prop.key);
        });
        Object.getOwnPropertyNames(val).forEach(key => {
            if (!Object.prototype.hasOwnProperty.call(props, key)) {
                result[key] = transform(val[key], additional, getProps, key);
            }
        });
        return result;
    }

    if (typ === "any") return val;
    if (typ === null) {
        if (val === null) return val;
        return invalidValue(typ, val);
    }
    if (typ === false) return invalidValue(typ, val);
    while (typeof typ === "object" && typ.ref !== undefined) {
        typ = typeMap[typ.ref];
    }
    if (Array.isArray(typ)) return transformEnum(typ, val);
    if (typeof typ === "object") {
        return typ.hasOwnProperty("unionMembers") ? transformUnion(typ.unionMembers, val)
            : typ.hasOwnProperty("arrayItems")    ? transformArray(typ.arrayItems, val)
            : typ.hasOwnProperty("props")         ? transformObject(getProps(typ), typ.additional, val)
            : invalidValue(typ, val);
    }
    // Numbers can be parsed by Date but shouldn't be.
    if (typ === Date && typeof val !== "number") return transformDate(val);
    return transformPrimitive(typ, val);
}

function cast<T>(val: any, typ: any): T {
    return transform(val, typ, jsonToJSProps);
}

function uncast<T>(val: T, typ: any): any {
    return transform(val, typ, jsToJSONProps);
}

function a(typ: any) {
    return { arrayItems: typ };
}

function u(...typs: any[]) {
    return { unionMembers: typs };
}

function o(props: any[], additional: any) {
    return { props, additional };
}

function m(additional: any) {
    return { props: [], additional };
}

function r(name: string) {
    return { ref: name };
}

const typeMap: any = {
    "Cfgu": o([
        { json: "contents", js: "contents", typ: "" },
        { json: "name", js: "name", typ: "" },
        { json: "path", js: "path", typ: "" },
        { json: "type", js: "type", typ: r("CfguType") },
    ], "any"),
    "CfguContents": o([
        { json: "default", js: "default", typ: u(undefined, "") },
        { json: "depends", js: "depends", typ: u(undefined, a("")) },
        { json: "description", js: "description", typ: u(undefined, "") },
        { json: "pattern", js: "pattern", typ: u(undefined, "") },
        { json: "required", js: "required", typ: u(undefined, true) },
        { json: "template", js: "template", typ: u(undefined, "") },
        { json: "type", js: "type", typ: r("ConfigSchemaType") },
    ], "any"),
    "Set": o([
        { json: "hierarchy", js: "hierarchy", typ: a("") },
        { json: "path", js: "path", typ: "" },
    ], "any"),
    "Store": o([
        { json: "scheme", js: "scheme", typ: "" },
        { json: "uid", js: "uid", typ: "" },
    ], "any"),
    "StoreQuery": o([
        { json: "key", js: "key", typ: "" },
        { json: "schema", js: "schema", typ: "" },
        { json: "set", js: "set", typ: "" },
    ], "any"),
    "StoreContents": o([
        { json: "key", js: "key", typ: "" },
        { json: "schema", js: "schema", typ: "" },
        { json: "set", js: "set", typ: "" },
        { json: "value", js: "value", typ: "" },
    ], "any"),
    "CfguType": [
        "json",
        "yaml",
    ],
    "ConfigSchemaType": [
        "Base64",
        "Boolean",
        "Color",
        "ConnectionString",
        "Country",
        "Currency",
        "Domain",
        "Email",
        "Hex",
        "IPv4",
        "IPv6",
        "LatLong",
        "Locale",
        "MD5",
        "MobilePhone",
        "Number",
        "RegEx",
        "SHA",
        "SemVer",
        "String",
        "URL",
        "UUID",
    ],
};

File: types-old/out.py
# To use this code, make sure you
#
#     import json
#
# and then, to convert JSON from a string, do
#
#     result = cfgu_type_from_dict(json.loads(json_string))
#     result = cfgu_from_dict(json.loads(json_string))
#     result = config_schema_type_from_dict(json.loads(json_string))
#     result = config_schema_from_dict(json.loads(json_string))
#     result = cfgu_contents_from_dict(json.loads(json_string))
#     result = set_from_dict(json.loads(json_string))
#     result = store_from_dict(json.loads(json_string))
#     result = store_query_element_from_dict(json.loads(json_string))
#     result = store_query_from_dict(json.loads(json_string))
#     result = config_from_dict(json.loads(json_string))
#     result = store_contents_from_dict(json.loads(json_string))
#     result = convert_from_dict(json.loads(json_string))

from enum import Enum
from typing import Any, Optional, List, Dict, TypeVar, Type, Callable, cast


T = TypeVar("T")
EnumT = TypeVar("EnumT", bound=Enum)


def from_str(x: Any) -> str:
    assert isinstance(x, str)
    return x


def to_enum(c: Type[EnumT], x: Any) -> EnumT:
    assert isinstance(x, c)
    return x.value


def from_none(x: Any) -> Any:
    assert x is None
    return x


def from_union(fs, x):
    for f in fs:
        try:
            return f(x)
        except:
            pass
    assert False


def from_list(f: Callable[[Any], T], x: Any) -> List[T]:
    assert isinstance(x, list)
    return [f(y) for y in x]


def from_bool(x: Any) -> bool:
    assert isinstance(x, bool)
    return x


def to_class(c: Type[T], x: Any) -> dict:
    assert isinstance(x, c)
    return cast(Any, x).to_dict()


def from_dict(f: Callable[[Any], T], x: Any) -> Dict[str, T]:
    assert isinstance(x, dict)
    return { k: f(v) for (k, v) in x.items() }


class CfguType(Enum):
    JSON = "json"
    YAML = "yaml"


class Cfgu:
    """A generic representation of a <schema>.cfgu.[json|yaml] file that contains ConfigSchema
    records ([<key>: string]: ConfigSchema)
    """
    contents: str
    name: str
    path: str
    type: CfguType

    def __init__(self, contents: str, name: str, path: str, type: CfguType) -> None:
        self.contents = contents
        self.name = name
        self.path = path
        self.type = type

    @staticmethod
    def from_dict(obj: Any) -> 'Cfgu':
        assert isinstance(obj, dict)
        contents = from_str(obj.get("contents"))
        name = from_str(obj.get("name"))
        path = from_str(obj.get("path"))
        type = CfguType(obj.get("type"))
        return Cfgu(contents, name, path, type)

    def to_dict(self) -> dict:
        result: dict = {}
        result["contents"] = from_str(self.contents)
        result["name"] = from_str(self.name)
        result["path"] = from_str(self.path)
        result["type"] = to_enum(CfguType, self.type)
        return result


class ConfigSchemaType(Enum):
    BASE64 = "Base64"
    BOOLEAN = "Boolean"
    COLOR = "Color"
    CONNECTION_STRING = "ConnectionString"
    COUNTRY = "Country"
    CURRENCY = "Currency"
    DOMAIN = "Domain"
    EMAIL = "Email"
    HEX = "Hex"
    I_PV4 = "IPv4"
    I_PV6 = "IPv6"
    LAT_LONG = "LatLong"
    LOCALE = "Locale"
    MD5 = "MD5"
    MOBILE_PHONE = "MobilePhone"
    NUMBER = "Number"
    REG_EX = "RegEx"
    SEM_VER = "SemVer"
    SHA = "SHA"
    STRING = "String"
    URL = "URL"
    UUID = "UUID"


class ConfigSchema:
    default: Optional[str]
    depends: Optional[List[str]]
    description: Optional[str]
    pattern: Optional[str]
    required: Optional[bool]
    template: Optional[str]
    type: ConfigSchemaType

    def __init__(self, default: Optional[str], depends: Optional[List[str]], description: Optional[str], pattern: Optional[str], required: Optional[bool], template: Optional[str], type: ConfigSchemaType) -> None:
        self.default = default
        self.depends = depends
        self.description = description
        self.pattern = pattern
        self.required = required
        self.template = template
        self.type = type

    @staticmethod
    def from_dict(obj: Any) -> 'ConfigSchema':
        assert isinstance(obj, dict)
        default = from_union([from_str, from_none], obj.get("default"))
        depends = from_union([lambda x: from_list(from_str, x), from_none], obj.get("depends"))
        description = from_union([from_str, from_none], obj.get("description"))
        pattern = from_union([from_str, from_none], obj.get("pattern"))
        required = from_union([from_bool, from_none], obj.get("required"))
        template = from_union([from_str, from_none], obj.get("template"))
        type = ConfigSchemaType(obj.get("type"))
        return ConfigSchema(default, depends, description, pattern, required, template, type)

    def to_dict(self) -> dict:
        result: dict = {}
        result["default"] = from_union([from_str, from_none], self.default)
        result["depends"] = from_union([lambda x: from_list(from_str, x), from_none], self.depends)
        result["description"] = from_union([from_str, from_none], self.description)
        result["pattern"] = from_union([from_str, from_none], self.pattern)
        result["required"] = from_union([from_bool, from_none], self.required)
        result["template"] = from_union([from_str, from_none], self.template)
        result["type"] = to_enum(ConfigSchemaType, self.type)
        return result


class Set:
    """A generic representation of a ConfigSet that contains a path hierarchy to differentiate
    Config values and enable inheritance
    """
    hierarchy: List[str]
    path: str

    def __init__(self, hierarchy: List[str], path: str) -> None:
        self.hierarchy = hierarchy
        self.path = path

    @staticmethod
    def from_dict(obj: Any) -> 'Set':
        assert isinstance(obj, dict)
        hierarchy = from_list(from_str, obj.get("hierarchy"))
        path = from_str(obj.get("path"))
        return Set(hierarchy, path)

    def to_dict(self) -> dict:
        result: dict = {}
        result["hierarchy"] = from_list(from_str, self.hierarchy)
        result["path"] = from_str(self.path)
        return result


class Store:
    """A generic representation of a ConfigStore that contains Config records (Config[])"""
    scheme: str
    uid: str

    def __init__(self, scheme: str, uid: str) -> None:
        self.scheme = scheme
        self.uid = uid

    @staticmethod
    def from_dict(obj: Any) -> 'Store':
        assert isinstance(obj, dict)
        scheme = from_str(obj.get("scheme"))
        uid = from_str(obj.get("uid"))
        return Store(scheme, uid)

    def to_dict(self) -> dict:
        result: dict = {}
        result["scheme"] = from_str(self.scheme)
        result["uid"] = from_str(self.uid)
        return result


class StoreQueryElement:
    key: str
    schema: str
    set: str

    def __init__(self, key: str, schema: str, set: str) -> None:
        self.key = key
        self.schema = schema
        self.set = set

    @staticmethod
    def from_dict(obj: Any) -> 'StoreQueryElement':
        assert isinstance(obj, dict)
        key = from_str(obj.get("key"))
        schema = from_str(obj.get("schema"))
        set = from_str(obj.get("set"))
        return StoreQueryElement(key, schema, set)

    def to_dict(self) -> dict:
        result: dict = {}
        result["key"] = from_str(self.key)
        result["schema"] = from_str(self.schema)
        result["set"] = from_str(self.set)
        return result


class Config:
    """A generic representation of a software configuration, aka Config"""
    key: str
    schema: str
    set: str
    value: str

    def __init__(self, key: str, schema: str, set: str, value: str) -> None:
        self.key = key
        self.schema = schema
        self.set = set
        self.value = value

    @staticmethod
    def from_dict(obj: Any) -> 'Config':
        assert isinstance(obj, dict)
        key = from_str(obj.get("key"))
        schema = from_str(obj.get("schema"))
        set = from_str(obj.get("set"))
        value = from_str(obj.get("value"))
        return Config(key, schema, set, value)

    def to_dict(self) -> dict:
        result: dict = {}
        result["key"] = from_str(self.key)
        result["schema"] = from_str(self.schema)
        result["set"] = from_str(self.set)
        result["value"] = from_str(self.value)
        return result


def cfgu_type_from_dict(s: Any) -> CfguType:
    return CfguType(s)


def cfgu_type_to_dict(x: CfguType) -> Any:
    return to_enum(CfguType, x)


def cfgu_from_dict(s: Any) -> Cfgu:
    return Cfgu.from_dict(s)


def cfgu_to_dict(x: Cfgu) -> Any:
    return to_class(Cfgu, x)


def config_schema_type_from_dict(s: Any) -> ConfigSchemaType:
    return ConfigSchemaType(s)


def config_schema_type_to_dict(x: ConfigSchemaType) -> Any:
    return to_enum(ConfigSchemaType, x)


def config_schema_from_dict(s: Any) -> ConfigSchema:
    return ConfigSchema.from_dict(s)


def config_schema_to_dict(x: ConfigSchema) -> Any:
    return to_class(ConfigSchema, x)


def cfgu_contents_from_dict(s: Any) -> Dict[str, ConfigSchema]:
    return from_dict(ConfigSchema.from_dict, s)


def cfgu_contents_to_dict(x: Dict[str, ConfigSchema]) -> Any:
    return from_dict(lambda x: to_class(ConfigSchema, x), x)


def set_from_dict(s: Any) -> Set:
    return Set.from_dict(s)


def set_to_dict(x: Set) -> Any:
    return to_class(Set, x)


def store_from_dict(s: Any) -> Store:
    return Store.from_dict(s)


def store_to_dict(x: Store) -> Any:
    return to_class(Store, x)


def store_query_element_from_dict(s: Any) -> StoreQueryElement:
    return StoreQueryElement.from_dict(s)


def store_query_element_to_dict(x: StoreQueryElement) -> Any:
    return to_class(StoreQueryElement, x)


def store_query_from_dict(s: Any) -> List[StoreQueryElement]:
    return from_list(StoreQueryElement.from_dict, s)


def store_query_to_dict(x: List[StoreQueryElement]) -> Any:
    return from_list(lambda x: to_class(StoreQueryElement, x), x)


def config_from_dict(s: Any) -> Config:
    return Config.from_dict(s)


def config_to_dict(x: Config) -> Any:
    return to_class(Config, x)


def store_contents_from_dict(s: Any) -> List[Config]:
    return from_list(Config.from_dict, s)


def store_contents_to_dict(x: List[Config]) -> Any:
    return from_list(lambda x: to_class(Config, x), x)


def convert_from_dict(s: Any) -> Dict[str, Any]:
    return from_dict(lambda x: x, s)


def convert_to_dict(x: Dict[str, Any]) -> Any:
    return from_dict(lambda x: x, x)
